tsl/src/fdw/modify_exec.c

```diff
@@ -393,6 +395,46 @@ response_type(AttConvInMetadata *att_conv_metadata)
 	return att_conv_metadata == NULL || att_conv_metadata->binary ? FORMAT_BINARY : FORMAT_TEXT;
 }
 
+static void
+fdw_chunk_update_stale_metadata(TsFdwModifyState *fmstate)
+{
+	List *all_data_nodes;
+	Relation rel = fmstate->rel;
+	int32 chunk_id = ts_chunk_get_id_by_relid(rel->rd_id);
+	Assert(chunk_id != 0);
+
+	all_data_nodes = ts_chunk_data_node_scan_by_chunk_id(chunk_id, CurrentMemoryContext);
+	Assert(list_length(all_data_nodes) >= fmstate->num_data_nodes);
+
+	if (list_length(all_data_nodes) > fmstate->num_data_nodes)
+	{
+		Chunk *chunk = ts_chunk_get_by_relid(rel->rd_id, true);
+		/* get filtered list */
+		List *serveroids = get_chunk_data_nodes(rel->rd_id);
+		ListCell *lc;
+		Assert(list_length(serveroids) == fmstate->num_data_nodes);
+
+		foreach (lc, all_data_nodes)
+		{
+			ChunkDataNode *cdn = lfirst(lc);
+			/*
+			 * check if this DN is a part of serveroids. If not
+			 * found in serveroids, then we need to remove this
+			 * chunk id to node name mapping and also update the primary
+			 * foreign server if necessary. It's possible that this metadata
+			 * might have been already cleared earlier but we have no way of
+			 * knowing that here.
+			 */
+			if (!list_member_oid(serveroids, cdn->foreign_server_oid))
+			{
+				chunk_update_foreign_server_if_needed(chunk, cdn->foreign_server_oid, false);
+				ts_chunk_data_node_delete_by_chunk_id_and_node_name(cdn->fd.chunk_id,
+																	NameStr(cdn->fd.node_name));
+			}
+		}
+	}
```

Comment:
This code looks the same as `chunk_update_stale_metadata` in `tsl/src/chunk.c`. Can we break that function up so that we can put this part of the code in a function instead of duplicating the code?